File: python/tests/test_ansatz.py
"""Comprehensive tests for the Ansatz class serialization in QDK/Chemistry."""

# --------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE.txt in the project root for license information.
# --------------------------------------------------------------------------------------------

import json
import pickle

import numpy as np
import pytest

from qdk_chemistry import algorithms
from qdk_chemistry.data import (
    Ansatz,
    CanonicalFourCenterHamiltonianContainer,
    CasWavefunctionContainer,
    Configuration,
    Hamiltonian,
    Orbitals,
    Structure,
    Wavefunction,
)

from .reference_tolerances import (
    float_comparison_absolute_tolerance,
    float_comparison_relative_tolerance,
    scf_energy_tolerance,
)
from .test_helpers import create_test_basis_set

try:
    import pyscf  # noqa: F401

    PYSCF_AVAILABLE = True
except ImportError:
    PYSCF_AVAILABLE = False


class TestAnsatzSerialization:
    """Test ansatz serialization and deserialization."""

    @pytest.fixture
    def basic_orbitals(self):
        """Create basic orbitals for testing."""
        coeffs = np.array([[0.9, 0.1], [0.1, -0.9], [0.0, 0.0]])
        basis_set = create_test_basis_set(3, "test-ansatz")
        return Orbitals(coeffs, None, None, basis_set)

    @pytest.fixture
    def test_structure(self):
        """Create a test structure (H2 molecule)."""
        positions = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.4]])
        elements = ["H", "H"]
        return Structure(positions, elements)

    @pytest.fixture
    def test_basis_set(self, test_structure):
        """Create a test basis set."""
        return create_test_basis_set(2, "test-ansatz-basis", test_structure)

    @pytest.fixture
    def test_wavefunction(self, basic_orbitals):
        """Create a test wavefunction."""
        det1 = Configuration("20")
        coeffs = np.array([1.0])  # Single determinant with coefficient 1.0

        container = CasWavefunctionContainer(coeffs, [det1], basic_orbitals)
        return Wavefunction(container)

    @pytest.fixture
    def test_ansatz(self, test_wavefunction):
        """Create a test ansatz."""
        # Create a hamiltonian using the same orbitals as the wavefunction
        orbitals = test_wavefunction.get_orbitals()
        num_orbitals = orbitals.get_num_molecular_orbitals()

        # Create simple one and two body integrals
        one_body = np.eye(num_orbitals)
        two_body = np.zeros(num_orbitals**4)
        fock = np.eye(0)

        test_hamiltonian = Hamiltonian(CanonicalFourCenterHamiltonianContainer(one_body, two_body, orbitals, 0.0, fock))
        return Ansatz(test_hamiltonian, test_wavefunction)

    def test_json_serialization(self, test_ansatz):
        """Test JSON serialization for ansatz."""
        # Test to_json and from_json
        json_str = test_ansatz.to_json()
        json_data = json.loads(json_str)

        # Verify essential fields
        assert "wavefunction" in json_data
        assert "hamiltonian" in json_data

        # Test round-trip serialization
        ansatz_reconstructed = Ansatz.from_json(json_str)

        # Verify nested objects are preserved
        assert ansatz_reconstructed.get_wavefunction() is not None
        assert ansatz_reconstructed.get_hamiltonian() is not None

        # Verify wavefunction properties
        orig_wf = test_ansatz.get_wavefunction()
        recon_wf = ansatz_reconstructed.get_wavefunction()
        assert recon_wf.size() == orig_wf.size()
        assert np.isclose(
            recon_wf.norm(),
            orig_wf.norm(),
            rtol=float_comparison_relative_tolerance,
            atol=float_comparison_absolute_tolerance,
        )

        # Verify hamiltonian properties
        orig_ham = test_ansatz.get_hamiltonian()
        recon_ham = ansatz_reconstructed.get_hamiltonian()
        assert recon_ham.get_core_energy() == orig_ham.get_core_energy()
        assert recon_ham.has_one_body_integrals() == orig_ham.has_one_body_integrals()
        assert recon_ham.has_two_body_integrals() == orig_ham.has_two_body_integrals()

    def test_hdf5_serialization(self, test_ansatz, tmp_path):
        """Test HDF5 serialization for ansatz."""
        filename = tmp_path / "test_ansatz.ansatz.h5"

        # Save to HDF5 file
        test_ansatz.to_hdf5_file(str(filename))

        # Load from HDF5 file
        ansatz_reconstructed = Ansatz.from_hdf5_file(str(filename))

        # Verify nested objects are preserved
        assert ansatz_reconstructed.get_wavefunction() is not None
        assert ansatz_reconstructed.get_hamiltonian() is not None

        # Verify wavefunction properties
        orig_wf = test_ansatz.get_wavefunction()
        recon_wf = ansatz_reconstructed.get_wavefunction()
        assert recon_wf.size() == orig_wf.size()
        assert np.isclose(
            recon_wf.norm(),
            orig_wf.norm(),
            rtol=float_comparison_relative_tolerance,
            atol=float_comparison_absolute_tolerance,
        )

        # Verify hamiltonian properties
        orig_ham = test_ansatz.get_hamiltonian()
        recon_ham = ansatz_reconstructed.get_hamiltonian()
        assert recon_ham.get_core_energy() == orig_ham.get_core_energy()
        assert recon_ham.has_one_body_integrals() == orig_ham.has_one_body_integrals()
        assert recon_ham.has_two_body_integrals() == orig_ham.has_two_body_integrals()

    def test_json_file_io(self, test_ansatz, tmp_path):
        """Test JSON file I/O."""
        filename = tmp_path / "test_ansatz.ansatz.json"

        # Save to JSON file
        test_ansatz.to_json_file(str(filename))

        # Load from JSON file
        ansatz_reconstructed = Ansatz.from_json_file(str(filename))

        # Verify nested objects are preserved
        assert ansatz_reconstructed.get_wavefunction() is not None
        assert ansatz_reconstructed.get_hamiltonian() is not None

    def test_generic_file_io(self, test_ansatz, tmp_path):
        """Test generic file I/O with different formats."""
        json_filename = tmp_path / "test_ansatz_generic.ansatz.json"
        hdf5_filename = tmp_path / "test_ansatz_generic.ansatz.h5"

        # Test JSON format
        test_ansatz.to_file(str(json_filename), "json")
        ansatz_json = Ansatz.from_file(str(json_filename), "json")
        assert ansatz_json.get_wavefunction() is not None

        # Test HDF5 format
        test_ansatz.to_file(str(hdf5_filename), "hdf5")
        ansatz_hdf5 = Ansatz.from_file(str(hdf5_filename), "hdf5")
        assert ansatz_hdf5.get_wavefunction() is not None

        # Test invalid format
        with pytest.raises(RuntimeError, match="Unsupported file type"):
            test_ansatz.to_file(str(tmp_path / "test.ansatz.xyz"), "xyz")

        with pytest.raises(RuntimeError, match="Unsupported file type"):
            Ansatz.from_file(str(tmp_path / "test.ansatz.xyz"), "xyz")

    def test_error_handling(self):
        """Test error handling for malformed data."""
        # Test malformed JSON string
        bad_json_str = json.dumps({"wavefunction": "invalid"})

        with pytest.raises(RuntimeError):
            Ansatz.from_json(bad_json_str)

        # Test non-existent files
        with pytest.raises(RuntimeError):
            Ansatz.from_json_file("non_existent.ansatz.json")

        with pytest.raises(RuntimeError):
            Ansatz.from_hdf5_file("non_existent.ansatz.h5")

    def test_nested_serialization_consistency(self, test_ansatz):
        """Test that nested objects maintain consistency through serialization."""
        # Serialize ansatz
        json_str = test_ansatz.to_json()
        ansatz_reconstructed = Ansatz.from_json(json_str)

        # Verify that the orbitals in the wavefunction and hamiltonian are consistent
        recon_wf = ansatz_reconstructed.get_wavefunction()
        recon_ham = ansatz_reconstructed.get_hamiltonian()

        recon_wf_orbitals = recon_wf.get_orbitals()
        recon_ham_orbitals = recon_ham.get_orbitals()

        # They should have the same basis set properties
        wf_basis = recon_wf_orbitals.get_basis_set()
        ham_basis = recon_ham_orbitals.get_basis_set()

        assert wf_basis.get_num_atomic_orbitals() == ham_basis.get_num_atomic_orbitals()
        assert wf_basis.get_name() == ham_basis.get_name()

    def test_repr_method(self, test_ansatz):
        """Test that __repr__ returns the summary."""
        repr_str = repr(test_ansatz)
        summary_str = test_ansatz.get_summary()
        assert repr_str == summary_str
        assert "Ansatz" in repr_str

    def test_str_method(self, test_ansatz):
        """Test that __str__ returns the summary."""
        str_str = str(test_ansatz)
        summary_str = test_ansatz.get_summary()
        assert str_str == summary_str
        assert "Ansatz" in str_str

    def test_pickling_ansatz(self, test_ansatz):
        """Test that Ansatz can be pickled and unpickled correctly."""
        # Test pickling round-trip
        pickled_data = pickle.dumps(test_ansatz)
        ansatz_restored = pickle.loads(pickled_data)

        # Verify core properties
        assert ansatz_restored.has_hamiltonian() == test_ansatz.has_hamiltonian()
        assert ansatz_restored.has_wavefunction() == test_ansatz.has_wavefunction()
        assert ansatz_restored.has_orbitals() == test_ansatz.has_orbitals()

        # Verify hamiltonian data
        if test_ansatz.has_hamiltonian():
            orig_ham = test_ansatz.get_hamiltonian()
            restored_ham = ansatz_restored.get_hamiltonian()
            assert orig_ham.get_core_energy() == restored_ham.get_core_energy()

            if orig_ham.has_one_body_integrals():
                assert np.array_equal(orig_ham.get_one_body_integrals()[0], restored_ham.get_one_body_integrals()[0])
                assert np.array_equal(orig_ham.get_one_body_integrals()[1], restored_ham.get_one_body_integrals()[1])

        # Verify wavefunction data
        if test_ansatz.has_wavefunction():
            orig_wf = test_ansatz.get_wavefunction()
            restored_wf = ansatz_restored.get_wavefunction()
            assert orig_wf.size() == restored_wf.size()
            assert np.isclose(
                orig_wf.norm(),
                restored_wf.norm(),
                rtol=float_comparison_relative_tolerance,
                atol=float_comparison_absolute_tolerance,
            )

            # Verify determinants
            orig_dets = orig_wf.get_active_determinants()
            restored_dets = restored_wf.get_active_determinants()
            assert len(orig_dets) == len(restored_dets)

            # Verify coefficients for each determinant
            for det in orig_dets:
                orig_coeff = orig_wf.get_coefficient(det)
                restored_coeff = restored_wf.get_coefficient(det)
                if isinstance(orig_coeff, complex):
                    assert np.isclose(
                        orig_coeff,
                        restored_coeff,
                        rtol=float_comparison_relative_tolerance,
                        atol=float_comparison_absolute_tolerance,
                    )
                else:
                    assert np.isclose(
                        orig_coeff,
                        restored_coeff,
                        rtol=float_comparison_relative_tolerance,
                        atol=float_comparison_absolute_tolerance,
                    )

        # Verify orbital consistency
        if test_ansatz.has_orbitals():
            orig_orbs = test_ansatz.get_orbitals()
            restored_orbs = ansatz_restored.get_orbitals()
            assert orig_orbs.get_num_molecular_orbitals() == restored_orbs.get_num_molecular_orbitals()
            assert np.array_equal(orig_orbs.get_coefficients(), restored_orbs.get_coefficients())

    def test_restricted_closed_shell_energy(self):
        """Test the energy evaluation for a restricted closed-shell system."""
        mol = Structure([[0.0, 0.0, 0.0], [0.0, 0.0, 4.0]], ["N", "N"])
        # get scf energy and wfn
        scf = algorithms.create("scf_solver")
        e_scf, hf_wfn = scf.run(mol, 0, 1, "cc-pvdz")
        # get hamiltonian
        h_ctor = algorithms.create("hamiltonian_constructor")
        hamiltonian = h_ctor.run(hf_wfn.get_orbitals())
        # get ansatz
        ansatz = Ansatz(hamiltonian, hf_wfn)
        e_rhf = ansatz.calculate_energy()
        # energy from ansatz should reproduce scf energy
        assert np.isclose(e_rhf, e_scf, rtol=float_comparison_relative_tolerance, atol=scf_energy_tolerance)

    @pytest.mark.skipif(not PYSCF_AVAILABLE, reason="pyscf not available")
    def test_restricted_open_shell_energy(self):
        """Test the energy evaluation for a restricted open-shell system."""
        try:
            import qdk_chemistry.plugins.pyscf as pyscf_plugin  # noqa: PLC0415

            pyscf_plugin.load()
        except ImportError:
            pytest.skip("pyscf not available, skipping O2 triplet Ansatz test")

        mol = Structure([[0.0, 0.0, 0.0], [0.0, 0.0, 4.0]], ["O", "O"])
        # get scf energy and wfn
        scf = algorithms.create("scf_solver", "pyscf")
        scf.settings().set("scf_type", "restricted")
        e_scf, hf_wfn = scf.run(mol, 0, 3, "cc-pvdz")
        # get hamiltonian
        h_ctor = algorithms.create("hamiltonian_constructor")
        hamiltonian = h_ctor.run(hf_wfn.get_orbitals())
        # get ansatz
        ansatz = Ansatz(hamiltonian, hf_wfn)
        e_rohf = ansatz.calculate_energy()
        # energy from ansatz should reproduce scf energy
        assert np.isclose(e_rohf, e_scf, rtol=float_comparison_relative_tolerance, atol=scf_energy_tolerance)


def test_ansatz_data_type_name():
    """Test that Ansatz has the correct _data_type_name class attribute."""
    assert hasattr(Ansatz, "_data_type_name")
    assert Ansatz._data_type_name == "ansatz"

File: cpp/tests/test_ansatz.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.txt in the project root for
// license information.

#include <gtest/gtest.h>

#include <cstdio>
#include <memory>
#include <nlohmann/json.hpp>
#include <qdk/chemistry/algorithms/active_space.hpp>
#include <qdk/chemistry/algorithms/hamiltonian.hpp>
#include <qdk/chemistry/algorithms/mc.hpp>
#include <qdk/chemistry/algorithms/scf.hpp>
#include <qdk/chemistry/constants.hpp>
#include <qdk/chemistry/data/ansatz.hpp>
#include <qdk/chemistry/data/hamiltonian.hpp>
#include <qdk/chemistry/data/hamiltonian_containers/canonical_four_center.hpp>
#include <qdk/chemistry/data/wavefunction.hpp>
#include <qdk/chemistry/data/wavefunction_containers/cas.hpp>
#include <qdk/chemistry/data/wavefunction_containers/sd.hpp>
#include <stdexcept>

#include "ut_common.hpp"

using namespace qdk::chemistry::data;
using namespace qdk::chemistry::algorithms;

class AnsatzSerializationTest : public ::testing::Test {
 protected:
  void SetUp() override {
    // Create test orbitals
    orbitals = testing::create_test_orbitals();

    // Create test wavefunction
    Eigen::VectorXd coeffs(2);
    coeffs << 0.8, 0.6;

    Wavefunction::DeterminantVector dets = {Configuration("200000"),
                                            Configuration("ud0000")};

    auto wf_container =
        std::make_unique<CasWavefunctionContainer>(coeffs, dets, orbitals);
    wavefunction = std::make_shared<Wavefunction>(std::move(wf_container));

    // Create test structure
    structure = testing::create_water_structure();

    // Create test hamiltonian with proper parameters (following
    // test_hamiltonian.cpp pattern)
    Eigen::MatrixXd one_body = Eigen::MatrixXd::Identity(2, 2);
    one_body(0, 1) = 0.5;
    one_body(1, 0) = 0.5;

    Eigen::VectorXd two_body =
        2 * Eigen::VectorXd::Ones(16);  // 2^4 = 16 for 2 orbitals
    double core_energy = 1.5;
    Eigen::MatrixXd inactive_fock = Eigen::MatrixXd::Zero(0, 0);

    hamiltonian = std::make_shared<Hamiltonian>(
        std::make_unique<CanonicalFourCenterHamiltonianContainer>(
            one_body, two_body, orbitals, core_energy, inactive_fock));

    // Create test ansatz
    ansatz = std::make_shared<Ansatz>(hamiltonian, wavefunction);
  }

  std::shared_ptr<Orbitals> orbitals;
  std::shared_ptr<Wavefunction> wavefunction;
  std::shared_ptr<Structure> structure;
  std::shared_ptr<Hamiltonian> hamiltonian;
  std::shared_ptr<Ansatz> ansatz;
};

TEST_F(AnsatzSerializationTest, JSONSerialization) {
  // Test that JSON serialization methods exist and can be called
  nlohmann::json j;
  EXPECT_NO_THROW(j = ansatz->to_json());

  // Verify essential fields are present
  EXPECT_TRUE(j.contains("wavefunction"));
  EXPECT_TRUE(j.contains("hamiltonian"));

  // Test that from_json method exists (but might not work correctly yet)
  // Just verify the API exists without requiring it to work perfectly
  EXPECT_NO_THROW(Ansatz::from_json(j));
}

TEST_F(AnsatzSerializationTest, HDF5Serialization) {
  // Test that HDF5 serialization methods exist and can be called
  std::string filename = "test_ansatz.ansatz.h5";

  // Test that the methods exist without requiring them to work perfectly
  EXPECT_NO_THROW(ansatz->to_hdf5_file(filename));

  // Test that from_hdf5_file method exists
  EXPECT_NO_THROW(Ansatz::from_hdf5_file(filename));

  // Clean up
  std::remove(filename.c_str());
}

TEST_F(AnsatzSerializationTest, JSONFileIO) {
  // Test JSON file I/O
  std::string filename = "test_ansatz.ansatz.json";

  // Save to JSON file
  ansatz->to_json_file(filename);

  // Load from JSON file
  auto ansatz_reconstructed = Ansatz::from_json_file(filename);
  EXPECT_NE(ansatz_reconstructed, nullptr);

  // Verify nested objects are preserved
  EXPECT_NE(ansatz_reconstructed->get_wavefunction(), nullptr);
  EXPECT_NE(ansatz_reconstructed->get_hamiltonian(), nullptr);
  EXPECT_NE(ansatz_reconstructed->get_orbitals(), nullptr);

  // Clean up
  std::remove(filename.c_str());
}

TEST_F(AnsatzSerializationTest, GenericFileIO) {
  // Test generic file I/O with different formats
  std::string json_filename = "test_ansatz_generic.ansatz.json";
  std::string hdf5_filename = "test_ansatz_generic.ansatz.h5";

  // Test JSON format
  ansatz->to_file(json_filename, "json");
  auto ansatz_json = Ansatz::from_file(json_filename, "json");
  EXPECT_NE(ansatz_json, nullptr);
  EXPECT_NE(ansatz_json->get_wavefunction(), nullptr);

  // Test HDF5 format
  ansatz->to_file(hdf5_filename, "hdf5");
  auto ansatz_hdf5 = Ansatz::from_file(hdf5_filename, "hdf5");
  EXPECT_NE(ansatz_hdf5, nullptr);
  EXPECT_NE(ansatz_hdf5->get_wavefunction(), nullptr);

  // Test invalid format
  EXPECT_THROW(ansatz->to_file("test.ansatz.xyz", "xyz"), std::runtime_error);
  EXPECT_THROW(Ansatz::from_file("test.ansatz.xyz", "xyz"), std::runtime_error);

  // Clean up
  std::remove(json_filename.c_str());
  std::remove(hdf5_filename.c_str());
}

TEST_F(AnsatzSerializationTest, ErrorHandling) {
  // Test error handling for malformed JSON
  nlohmann::json bad_json;
  bad_json["wavefunction"] = "invalid";

  EXPECT_THROW(Ansatz::from_json(bad_json), std::runtime_error);

  // Test error handling for non-existent files
  EXPECT_THROW(Ansatz::from_json_file("non_existent.ansatz.json"),
               std::runtime_error);
  EXPECT_THROW(Ansatz::from_hdf5_file("non_existent.ansatz.h5"),
               std::runtime_error);
}

class AnsatzEnergyCalculationTest : public ::testing::Test {
 protected:
  void SetUp() override {}
};

TEST_F(AnsatzSerializationTest, TestDataTypeName) {
  // test the data_type_name property
  EXPECT_EQ(ansatz->get_data_type_name(), "ansatz");
}

TEST_F(AnsatzEnergyCalculationTest, N2SingletCAS_6e6o) {
  // N2 structure
  auto structure = testing::create_stretched_n2_structure();

  // get wavefunction
  auto scf = ScfSolverFactory::create();
  const auto& [E_scf, wfn_scf] = scf->run(structure, 0, 1, "def2-svp");

  // // get full hamiltonian
  auto hamil_ctor = HamiltonianConstructorFactory::create();
  auto hamiltonian_hf = hamil_ctor->run(wfn_scf->get_orbitals());

  // get ansatz and energy
  auto ansatz_hf = Ansatz(hamiltonian_hf, wfn_scf);
  double energy_hf = ansatz_hf.calculate_energy();

  EXPECT_NEAR(energy_hf, E_scf, testing::scf_energy_tolerance);

  // select active space
  auto active_space = ActiveSpaceSelectorFactory::create("qdk_valence");
  active_space->settings().set("num_active_electrons", 6);
  active_space->settings().set("num_active_orbitals", 6);
  auto active_space_wfn = active_space->run(wfn_scf);

  // get hamiltonian
  auto hamiltonian_cas = hamil_ctor->run(active_space_wfn->get_orbitals());

  // get cas wavefunction
  auto mc_calc = MultiConfigurationCalculatorFactory::create("macis_cas");
  mc_calc->settings().set("calculate_two_rdm", true);
  mc_calc->settings().set("calculate_one_rdm", true);
  auto [E_cas, wfn_cas] = mc_calc->run(
      hamiltonian_cas, active_space_wfn->get_active_num_electrons().first,
      active_space_wfn->get_active_num_electrons().second);

  // get ansatz and energy
  auto ansatz = Ansatz(hamiltonian_cas, wfn_cas);
  double energy = ansatz.calculate_energy();

  // energy should match SCF energy
  EXPECT_NEAR(energy, E_cas, testing::scf_energy_tolerance);
}

TEST_F(AnsatzEnergyCalculationTest, O2TripletCAS_8e6o) {
  // O2 structure
  auto structure = testing::create_o2_structure();

  // get wavefunction
  auto scf = ScfSolverFactory::create();
  const auto& [E_scf, wfn_scf] = scf->run(structure, 0, 3, "def2-svp");

  // get full hamiltonian
  auto hamil_ctor = HamiltonianConstructorFactory::create();
  auto hamiltonian_hf = hamil_ctor->run(wfn_scf->get_orbitals());

  // get ansatz and energy
  auto ansatz_hf = Ansatz(hamiltonian_hf, wfn_scf);
  double energy_hf = ansatz_hf.calculate_energy();

  EXPECT_NEAR(energy_hf, E_scf, testing::scf_energy_tolerance);
}

// Test for creating Ansatz from separately loaded Hamiltonian and Wavefunction
// files
class AnsatzFromSeparateFilesTest : public ::testing::Test {
 protected:
  void SetUp() override {
    // Create H2 structure
    std::vector<Eigen::Vector3d> coords = {{0.0, 0.0, 0.0}, {0.0, 0.0, 1.4}};
    // Convert to Bohr
    for (auto& coord : coords) {
      coord *= qdk::chemistry::constants::angstrom_to_bohr;
    }
    std::vector<Element> elements = {Element::H, Element::H};
    structure = std::make_shared<Structure>(coords, elements);
  }

  std::shared_ptr<Structure> structure;
};

TEST_F(AnsatzFromSeparateFilesTest, CreateFromSeparatelyLoadedFiles) {
  // Run SCF to get wavefunction
  auto scf_solver = ScfSolverFactory::create();
  const auto& [E_scf, wfn] = scf_solver->run(structure, 0, 1, "sto-3g");

  // Build Hamiltonian from the same orbitals
  auto ham_constructor = HamiltonianConstructorFactory::create();
  auto ham = ham_constructor->run(wfn->get_orbitals());

  // Creating Ansatz from in-memory objects should work
  std::shared_ptr<Ansatz> ansatz_direct;
  EXPECT_NO_THROW(ansatz_direct = std::make_shared<Ansatz>(ham, wfn));
  EXPECT_NE(ansatz_direct, nullptr);

  // Create temporary files
  std::string wfn_path = "test_ansatz_wfn.wavefunction.json";
  std::string ham_path = "test_ansatz_ham.hamiltonian.json";
  std::string ansatz_path = "test_ansatz_unit.ansatz.json";

  // Save objects to separate files
  wfn->to_json_file(wfn_path);
  ham->to_json_file(ham_path);
  ansatz_direct->to_json_file(ansatz_path);

  // Loading Ansatz that was saved as a unit should work
  auto loaded_ansatz = Ansatz::from_json_file(ansatz_path);
  EXPECT_NE(loaded_ansatz, nullptr);
  EXPECT_NE(loaded_ansatz->get_hamiltonian(), nullptr);
  EXPECT_NE(loaded_ansatz->get_wavefunction(), nullptr);

  // Load Hamiltonian and Wavefunction separately
  auto loaded_wfn = Wavefunction::from_json_file(wfn_path);
  auto loaded_ham = Hamiltonian::from_json_file(ham_path);
  EXPECT_NE(loaded_wfn, nullptr);
  EXPECT_NE(loaded_ham, nullptr);

  // Verify orbitals are structurally equivalent
  auto orig_orbs = wfn->get_orbitals();
  auto loaded_wfn_orbs = loaded_wfn->get_orbitals();
  auto loaded_ham_orbs = loaded_ham->get_orbitals();

  EXPECT_EQ(orig_orbs->get_num_molecular_orbitals(),
            loaded_wfn_orbs->get_num_molecular_orbitals());
  EXPECT_EQ(orig_orbs->get_num_molecular_orbitals(),
            loaded_ham_orbs->get_num_molecular_orbitals());
  EXPECT_EQ(orig_orbs->is_restricted(), loaded_wfn_orbs->is_restricted());
  EXPECT_EQ(orig_orbs->is_restricted(), loaded_ham_orbs->is_restricted());

  // Creating Ansatz from separately loaded files should work
  std::shared_ptr<Ansatz> ansatz_loaded;
  EXPECT_NO_THROW(ansatz_loaded =
                      std::make_shared<Ansatz>(loaded_ham, loaded_wfn));
  EXPECT_NE(ansatz_loaded, nullptr);
  EXPECT_NE(ansatz_loaded->get_hamiltonian(), nullptr);
  EXPECT_NE(ansatz_loaded->get_wavefunction(), nullptr);

  // Mixed original + loaded should work
  std::shared_ptr<Ansatz> ansatz_mixed1;
  EXPECT_NO_THROW(ansatz_mixed1 = std::make_shared<Ansatz>(ham, loaded_wfn));
  EXPECT_NE(ansatz_mixed1, nullptr);

  std::shared_ptr<Ansatz> ansatz_mixed2;
  EXPECT_NO_THROW(ansatz_mixed2 = std::make_shared<Ansatz>(loaded_ham, wfn));
  EXPECT_NE(ansatz_mixed2, nullptr);

  // Clean up
  std::remove(wfn_path.c_str());
  std::remove(ham_path.c_str());
  std::remove(ansatz_path.c_str());
}

TEST_F(AnsatzFromSeparateFilesTest, CreateFromSeparatelyLoadedHDF5Files) {
  // Run SCF to get wavefunction
  auto scf_solver = ScfSolverFactory::create();
  const auto& [E_scf, wfn] = scf_solver->run(structure, 0, 1, "sto-3g");

  // Build Hamiltonian from the same orbitals
  auto ham_constructor = HamiltonianConstructorFactory::create();
  auto ham = ham_constructor->run(wfn->get_orbitals());

  // Create Ansatz from in-memory objects
  auto ansatz_direct = std::make_shared<Ansatz>(ham, wfn);

  // Create temporary files
  std::string wfn_path = "test_ansatz_wfn.wavefunction.h5";
  std::string ham_path = "test_ansatz_ham.hamiltonian.h5";
  std::string ansatz_path = "test_ansatz_unit.ansatz.h5";

  // Save objects to separate files
  wfn->to_hdf5_file(wfn_path);
  ham->to_hdf5_file(ham_path);
  ansatz_direct->to_hdf5_file(ansatz_path);

  // Load Hamiltonian and Wavefunction separately
  auto loaded_wfn = Wavefunction::from_hdf5_file(wfn_path);
  auto loaded_ham = Hamiltonian::from_hdf5_file(ham_path);

  // Creating Ansatz from separately loaded HDF5 files should work
  std::shared_ptr<Ansatz> ansatz_loaded;
  EXPECT_NO_THROW(ansatz_loaded =
                      std::make_shared<Ansatz>(loaded_ham, loaded_wfn));
  EXPECT_NE(ansatz_loaded, nullptr);

  // Clean up
  std::remove(wfn_path.c_str());
  std::remove(ham_path.c_str());
  std::remove(ansatz_path.c_str());
}

TEST_F(AnsatzFromSeparateFilesTest, DetectsOrbitalMismatch) {
  // Run SCF to get first wavefunction (H2)
  auto scf_solver = ScfSolverFactory::create();
  const auto& [E_scf, wfn] = scf_solver->run(structure, 0, 1, "sto-3g");

  // Build Hamiltonian from the first wavefunction's orbitals
  auto ham_constructor = HamiltonianConstructorFactory::create();
  auto ham = ham_constructor->run(wfn->get_orbitals());

  // Create a different structure (water) to get different orbitals
  auto water_structure = testing::create_water_structure();
  const auto& [E_water, wfn_water] =
      scf_solver->run(water_structure, 0, 1, "sto-3g");

  // Attempting to create Ansatz with mismatched orbitals should fail
  EXPECT_THROW(
      { auto ansatz = Ansatz(ham, wfn_water); }, std::invalid_argument);
}
